src/host/services/IMUService.cpp

```diff
@@ -26,38 +26,125 @@
+  _offsetRoll  = 0.0;
+  _offsetPitch = 0.0;
+
+  // until config for positioning is made, declare default mounting position KBox here
+  uint8_t axisConfig = 0b00001001;
+  uint8_t signConfig = 0b00000000;
+
   DEBUG("Initing BNO055");
-  if (!bno055.begin()) {
+  if (!bno055.begin(bno055.OPERATION_MODE_NDOF, axisConfig, signConfig)) {
     DEBUG("Error initializing BNO055");
   }
   else {
-    DEBUG("Success!");
+    if (recallCalibration())
+      DEBUG("Success!");
   }
 }
 
 void IMUService::loop() {
-  bno055.getCalibration(&sysCalib, &gyroCalib, &accelCalib, &magCalib);
+  bno055.getCalibration(&_sysCalib, &_gyroCalib, &_accelCalib, &_magCalib);
+  //DEBUG("Calib Sys: %i Accel: %i Gyro: %i Mag: %i", _sysCalib, _accelCalib, _gyroCalib, _magCalib);
 
   eulerAngles = bno055.getVector(Adafruit_BNO055::VECTOR_EULER);
-
-  //DEBUG("Calib Sys: %i Accel: %i Gyro: %i Mag: %i", sysCalib, accelCalib, gyroCalib, magCalib);
-  //DEBUG("Attitude roll: %f pitch: %f  Mag heading: %f", eulerAngles.z(), eulerAngles.y(), eulerAngles.x());
+  //DEBUG("Sensor Raw-datas: eulerAngles.z=%.3f eulerAngles.y=%.3f  eulerAngles.x=%.3f", eulerAngles.z(), eulerAngles.y(), eulerAngles.x());
 
   SKUpdateStatic<2> update;
   // Note: We could calculate yaw as the difference between the Magnetic
   // Heading and the Course Over Ground Magnetic.
 
-  /* if orientation == MOUNTED_ON_PORT_HULL */
-  double roll, pitch, heading;
-  roll = eulerAngles.z();
-  pitch = eulerAngles.y();
-  heading = fmod(eulerAngles.x() + 270, 360);
+  // TODO: different calcs according to mounting position
+  /* if orientation == MOUNTED_ON_STARBOARD_HULL */
+  _roll = SKDegToRad(eulerAngles.z());
+  _pitch = SKDegToRad(eulerAngles.y());
+  _heading = SKDegToRad(fmod(eulerAngles.x() + 270, 360));
 
-  if (sysCalib == 3) {
-    update.setNavigationAttitude(SKTypeAttitude(/* roll */ SKDegToRad(roll), /* pitch */ SKDegToRad(pitch), /* yaw */ 0));
-    update.setNavigationHeadingMagnetic(SKDegToRad(heading));
+  // update NavigationHeadingMagnetic if quality (= calibration value) is ok
+  if (_magCalib >= _config.calHdg) {
+    update.setNavigationHeadingMagnetic(_heading);
     _skHub.publish(update);
   }
+  // update NavigationAttitude if quality (= calibration value) is ok
+	if (_accelCalib >= _config.calHeelPitch && _gyroCalib >= _config.calHeelPitch) {
+    update.setNavigationAttitude(SKTypeAttitude(/* roll */ _roll + _offsetRoll, /* pitch */ _pitch + _offsetPitch, /* yaw */ 0));
+    //DEBUG("Heel = %.3f | Pitch = %.3f | Heading = %.3f", SKRadToDeg( _roll + _offsetRoll), SKRadToDeg( _pitch + _offsetPitch), SKRadToDeg(_heading));
+    _skHub.publish(update);
+  }
+}
+
+// get the actual values of calibration, heel & pitch (including offset) and heading for display
+// all angles in radians
+void IMUService::getLastValues(int &accelCalibration, double &pitch, double &roll, int &magCalibration, double &heading){
+  accelCalibration = _accelCalib;
+  pitch = _pitch + _offsetPitch;
```

Comment:
is it possible that the offset pushes the value beyond a theoretical maximum?

For example max pitch is +180, if the boat is pointing down and the offset is 5 degrees, we could have 179+5=184 which would be outside our range. Maybe we should apply some normalization here.

Reply:
The offset is meant for doing small corrections only, not for correcting mounting positions. E.g. after loading people, water, luggage.
If pitch is around 180 then you have (other) problems :-)
 If someone is sailing with heel 40° and wants to do have heel 0 on the display then this would be crazy.